### ISSUE TITLE
terraform test command does not honor TF_VAR_* environment variables when "module" is specified after 1.8.0

### ISSUE THREAD

**Issue by bminahan73** [hashicorp/terraform#35005]
### Terraform Version

```shell
Terraform v1.8.0
on linux_amd64
```

### Terraform Configuration Files

a dummy module, with a test, and a module used by the test:

main.tf:
```hcl
resource null_resource foo {}
```

foo/main.tf:
```hcl
variable "foo" {}
output "foo" {
    value = var.foo
}
```

foo.tftest.hcl:
```hcl
run "foo" {
    module {
        source = "./foo"
    }
}
```

### Debug Output

cannot post publicly.

### Expected Behavior

```bash
$ TF_VAR_foo=something terraform test
foo.tftest.hcl... in progress
  run "foo"... pass
foo.tftest.hcl... tearing down
foo.tftest.hcl... pass

Success! 1 passed, 0 failed.
```

### Actual Behavior

```bash
$ TF_VAR_foo=something terraform test
foo.tftest.hcl... in progress
  run "foo"... fail
╷
│ Error: Required variable not set
│ 
│   on foo/main.tf line 1:
│    1: variable "foo" {}
│ 
│ The variable "foo" is required, but is not set.
╵
foo.tftest.hcl... tearing down
foo.tftest.hcl... fail
Failure! 0 passed, 1 failed.
```

### Steps to Reproduce

```bash
TF_VAR_foo=something terraform test
```

### Additional Context

this does not happen on 1.7.x terraform versions. This also only seems to happen for tests with a custom module, ie:
```hcl
run {
  module {
    source = "./a-custom-module"
  }
}
```

**Comment by mjgoethe** [comment]
This appears to be an issue with the parsing behavior for environment variables and variables passed via command line for this specific scenario. I noticed that bools and integers still seem to work, but strings fail. Wrapping the string value with quotes gets rid of the error, but that's definitely not in line with normal parsing behavior.

**Comment by liamcervante** [comment]
Thanks @bminahan73, I can replicate this so I've marked it as confirmed.

It seems that we're not connecting the global variables (ie. environment variables, command line variables or variable file variables) up to the types defined within the alternate modules correctly. This means that Terraform doesn't know the correct type for your variable, and it is silently swallowing that error.

There's two errors in there. Firstly, it's not working out the type when it should, and secondly, it's not reporting the type error and simply stating the variable isn't defined when it is.

As pointed out by @mjgoethe, you can force the correct type by surrounding the input value in quotation marks which forces the type into a string and allows Terraform to parse it properly. This might work for you as a short-term workaround.

I will try and have a fix in place for this in time for v1.8.1. Thanks again for filing this.

**Comment by bminahan73** [comment]
Thank you for the fast fix!

**Comment by seblatre** [comment]
Is it available in 1.8.2?

**Comment by crw** [comment]
Yes, please see the release notes: https://github.com/hashicorp/terraform/blob/v1.8/CHANGELOG.md#182-april-24-2024
